### ISSUE TITLE
(Request) Select friend support by friend name and class besides servants/CEs 

### ISSUE THREAD

**Issue by muvluvalt** [29988122/Fate-Grand-Order_Lua#212]
It would be nice to have these 2 features, especially when one is going through the Main Quest and the classes differ for each node (meaning that the Servant/CE selected would vary immensely). The current option for this is just to set support selection to manual, but that kind of defeats the whole purpose of automation. Setting support selection to first is also not a viable option, because that would then select the NPC support instead of friend support.

Therefore, automatically selecting support by friend name, and by class, would help to mitigate this issue by selecting the desired support without having to take multiple Servant and CE screenshots.

In other words:

- In an archer node, user will be brought to the Lancer class section of the support list.

- In a mixed node (eg. Saber,Lancer), user will be brought to some random class section with class advantage over one class in that node (eg. Archer) of the support list.

In the latter case, selecting the Berserker class may be the better option since Archers are also weak to Lancers. The script doesn't switch classes (i.e from Archer to Berserker), which is one of the two features I'm requesting. 

P.S I'm not sure if this feature has already been implemented, so if it's already implemented, I will close this issue.

**Comment by TryBane** [comment]
I believe I wouldn't have an issue setting the script up to select a support by friend name, however you would still have to create a screenshot of that friends name. Otherwise we would literally need to create a library of images for every single letter of the alphabet AND any other language to use when you are selecting the support.

The reason we would need that, is because there is no limit to the characters people can use when creating their names. You are only limited in the number of characters.

As for the selection of the class, I believe that one might be a bit further than we would like to go. Because small UI differences between versions regarding their position, size, etc can cause incompatibility between versions. Example, say they introduce a new tab for supports. But now this tab needs to be squeezed in with the current line of them. This will either expand the size of the region needed to cover, which would be the ideal scenario, or the more likely instance would be them shrinking the size of all the other tab images to fit the new one. And this would require a replacement of all the current images. Not only that, but it would result in needing multiple copies of them to cover all versions, both the ones that have the new tab and the ones that dont. 

Thus is my reasoning for why I believe it would be very burdensome to implement and then maintain.

In conclusion, I could very easily implement a system that can select a servant purely based on name, but this would still require screenshots to be taken and adjusted by the user, such as yourself. But I may implement this anyways, because it sounds like something I would likely use as well. But the others would be difficult to maintain and excessively inflate the size the script takes up on anyone's device.

**Comment by muvluvalt** [comment]
If creating a function for selecting class is too troublesome, then I suppose you can just forget about it. The friend name function would be useful though; I'm fine with just creating a screenshot of a friend name.

**Comment by TryBane** [comment]
Alrighty, I'll begin working on it right now on my own personal fork. If you would like to test it a bit yourself, then I can inform you of when it my first iteration is completed so you can give it a try.

**Comment by muvluvalt** [comment]
Alright noted I'll give it a test once your first iteration is out. I'll post any error messages if they pop up.

**Comment by TryBane** [comment]
@muvluvalt First iteration finished! Please give it a try and provide any feedback you may have!

There is a new option in the FGO_XX_REGULAR.lua file for "FriendNames". Please place any file names of the image for your friend name in there! Don't have too much of the white background though, or else it'll mess up too much. An example of what you can do is below. Make a file like this the same way you would a servant!
![myFriend](https://user-images.githubusercontent.com/24114484/63558107-88b61780-c519-11e9-9521-784a95a7e70e.png)

I have already tested this on my Nox and it has worked as expected thus far. That being said, I always expect things to go wrong, so please inform me in this thread should something happen that shouldn't. Below is a link to the repo that will provide the functionality. I have more things I plan to implement shortly too in preparation for the new lotto events!

https://github.com/TryBane/Fate-Grand-Order_Lua/tree/Friend

**Comment by muvluvalt** [comment]
I got this error when I tried to start the script:

![image](https://user-images.githubusercontent.com/40586164/63569200-f1010980-c5aa-11e9-9e01-e4e419f9326a.png)

Any ideas why?

**Comment by TryBane** [comment]
Did you replace the FGO_XX_REGULAR.lua file?

**Comment by muvluvalt** [comment]
Yes I did. Here's a copy of my FGO_EN_REGULAR.lua file.
[FGO_EN_REGULAR.zip](https://github.com/29988122/Fate-Grand-Order_Lua/files/3532963/FGO_EN_REGULAR.zip)

**Comment by TryBane** [comment]
Oh no.... no no no XD

Don't comment out the other support options. That is more than likely what is breaking the script.

It won't try to access those if you have friend as the option you are running with

**Comment by muvluvalt** [comment]
Oh I see, my bad lol. By the way, are we required to change the dimensions of the screenshots? If not, by how much?

Because I've tried it with two different friend screenshots and the script doesn't seem to select either of them:

`--AutoSupportSelection
Support_SelectionMode = "friend"
Support_SwipesPerUpdate = 10
Support_MaxUpdates = 10
Support_FallbackTo = "manual"
Support_FriendsOnly = 1
Support_PreferredServants = "waver4.png, waver3.png, waver2.png, waver1.png"
Support_PreferredCEs = "*chaldea_lunchtime.png"
-- Support_PreferredServants = ""
-- Support_PreferredCEs = "*mona_lisa.png"
Support_FriendNames = "ponkan.png,cheesekid.png"`

![cheesekid](https://user-images.githubusercontent.com/40586164/63570560-412e9a80-c5b0-11e9-902a-f6a2639b208b.PNG)

![image](https://user-images.githubusercontent.com/40586164/63570589-560b2e00-c5b0-11e9-8956-f143c1320430.png)

![ponkan](https://user-images.githubusercontent.com/40586164/63570566-45f34e80-c5b0-11e9-9920-f5eb645047ea.PNG)

![image](https://user-images.githubusercontent.com/40586164/63570610-658a7700-c5b0-11e9-9332-a8018a574c44.png)

**Comment by TryBane** [comment]
You do the same thing as you would for Servant or CEs. So you take a screenshot of the full screen in the support selection menu, resize the whole image to 1280x720, crop out the name, and then save it and move it to the Support folder like you would for Servants and CEs

**Comment by muvluvalt** [comment]
Okay I've tested it on NA and JP and it's working fine on both. Thanks a lot!

**Comment by TryBane** [comment]
No problem! Good to know. Let me know if there are any other issues using it

**Comment by Rajvir8** [comment]
Just thought I'd ask if Friend works together with other Image recognition?

For example, I have created a screenshot for my friend Anon, and also put MLB Petal in the CE section.  If he moved his petal off of his Servant will it fail and switch to the default fallback option which for me is Manual? 

This is what I'd prefer just in case the friend I'm relying on moves their CE, currently I'd prefer to rely only on Preferred looking for MLB CE just in case someone moves it for now if it will always use the friend no matter what.

**Comment by TryBane** [comment]
No, that kind of combination is not setup currently. With the new format for getting a region setup though, it shouldn't be difficult to implement it. In fact, it would probably be able to work with both Servant AND CE.

My current use of the friend one though is basically just looking for friends I am in contact with regularly that have the Servant/CE I want and that I've asked them to have on for the duration of the Nerofest event's first and second rotations.

**Comment by Rajvir8** [comment]
Perfectly understandable, I'll look forward to when it does get implemented then, but for now I'll stick to preferred.  I did ask my friend not to move it, but they said several others may also need it and they haven't been playing enough to get two MLB Petals like I have sadly.

**Comment by TryBane** [comment]
Yeah, that's fine. I don't think it would be difficult to implement, but I do think it would be difficult to test while doing Nerofest currently and with most people using it for Nerofest as well.

**Comment by Rajvir8** [comment]
Completely understandable, I appreciate you adding it after Nerofest as well.  Thanks as always.
